.github/workflows/release.yml

```diff
@@ -0,0 +1,107 @@
+name: Release build
+
+on:
+  release:
+    types: [published]
+  workflow_dispatch:
+    inputs:
+      version:
+        description: "Test version"
+        required: true
+        default: "0.0.1"
+jobs:
+  build:
+    runs-on: windows-2022
+
+    defaults:
+      run:
+        shell: pwsh
+
+    env:
+      MCDUServerFileNameBase: "cj4-mcdu"
+
+    steps:
+      - uses: actions/checkout@v3
+
+      # Get the version of the GitHub release for use later. This is only valid
+      # when run in a release context
+      - name: Get GitHub release version
+        id: get_github_version
+        uses: battila7/get-version-action@v2
+        if: startsWith(github.ref, 'refs/tags/')
+      - name: Store GitHub release version
+        id: save_github_version
+        if: startsWith(github.ref, 'refs/tags/')
+        run: |
+          echo "version=${{ steps.get_github_version.outputs.version-without-v }}" >> $env:GITHUB_ENV
+          echo ${{ steps.get_github_version.outputs.version-without-v }} > ./version
+
+      # Get the version from the workflow input. This is only valid
+      # when run manually as a workflow.
+      - name: Store workflow version
+        id: save_workflow_version
+        if: startsWith(github.ref, 'refs/tags/') == false
+        run: |
+          echo "version=${{ github.event.inputs.version }}" >> $env:GITHUB_ENV
+          echo ${{ github.event.inputs.version }} > ./version
+
+      - name: Use Node.js 16.x
+        id: install_node
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16.x
+          cache: "npm"
+          cache-dependency-path: "src/app/package-lock.json"
+
+      - name: Set versions in code
+        id: set_versions
+        run: |
+          echo "Building with version ${{ env.version }}"
+          node ./updateversions.js
+        working-directory: src
+
+      - name: Setup node dependencies
+        id: install_dependencies
+        working-directory: src/app
+        run: npm ci
+
+      - name: Build client
+        id: build_client
+        working-directory: src/app
+        run: npm run build:client
+
+      - name: Build server
+        id: build_server
+        working-directory: src/app
+        run: npm run build:mcdu-server
+
+      - name: Build mod
+        id: build_mod
+        working-directory: src/mod
+        run: ./build.ps1
+
+      - name: Build release folder contents
+        id: build_release_folder
+        run: |
+          Move-Item "build/MCDU SERVER/${{ env.MCDUServerFileNameBase }}-server.exe" "build/MCDU SERVER/${{ env.MCDUServerFileNameBase }}-server-${{ env.version }}.exe"
+          Copy-Item readme.md build
+          Copy-Item version build
+
+      - name: Zip output
+        id: zip_output
+        working-directory: build
+        run: Compress-Archive ./* "../${{ env.MCDUServerFileNameBase }}-${{ env.version }}.zip"
+
```

Comment:
This looks like what `src/build.ps1` file was doing.    Why break it out like this instead of just running that script?

My preference is to keep as much complexity as possible in files that can be run locally, and keep the CI scripts simple invocations of those scripts.

Reply:
Mostly because I was reusing components that I understand from other github release workflows I've done, like pulling the version number from the git tag/workflow invocation, and that lead me to go down the path of doing the rest via GitHub actions like I've done in other projects. It also made it easier to debug paths since I could get each step working one at a time.

Lines 68-93 could be replaced with the old build.ps1 I think, if `npm install` changes to `npm ci`. I can give that a try if you like and report back the results.

Reply:
Yeah, go ahead and make that change.   `build.ps1` could take the version as an optional parameter so we get it from the action if available or have a sensible default if running locally.

Reply:
I'm already writing it to the `./version` file earlier in the workflow so the script can just keep taking it that way unchanged. I had to do that so the `mod/build.ps1` script would "just work".

Giving the build.ps1 approach a try now, standby.

Reply:
I think `node updateversions.js` was in your previous build but is missing from this one.

The fact that the version can come from 3 places (/version, the xml/json files, or the git tag) bothers me a bit.  Maybe for releases, `updateversions.js` will fail the build if the version in the files doesn't match the version in the tag?

Not sure there's a good solution here, looking for the least bad one.

Reply:
Yep, I forgot it isn't handled by build.ps1. Oops. It's added back now. https://github.com/neilenns/cj4-mcdu/actions/runs/3752742936/jobs/6375249831

The way I've handled this in past repos is to have GitHub's release tab be the place to set the version. It's once place that's done at the time you truly are cutting a release and hard to screw up. Then I let actions/scripts take care of writing it to all the correct places, similar to what you'd already done. It also ensures the version matches the github tag. It's technically possible to check those changed files in as well as part of the release script but I've never personally cared for that.

Any time I've had projects where I have to remember to put the version in a file and check it in, I forget and cut a release with the old version🤦‍♂️

I was trying not to change your existing scripts too much so I left the `./version` file around instead of just going to environment variables in the scripts, especially since you were including it in the downloadable zip file.

Reply:
Thinking about this a bit more, if you're willing to leave the `version` file out of the zip this could simplify down a bit and [follow a pattern I've used elsewhere](https://github.dev/MobiFlight/MobiFlight-FirmwareSource/blob/b1e0ddb243729b2bddb13c09a1d92a124916fc0d/get_version.py#L4-L11).

1. Version comes from github during releases via an environment variable.
2. One of the scripts tests for the presence of the version environment variable, and if missing defaults to something like 0.0.1.
3. The scripts update to use the environment variable instead of the `version` file as the source for the version number.